### PR TITLE
Config redis and smtp using env vars.

### LINKED ISSUE #194: Can I setup redis via env var ?
Hello

I'ld like to setup a redis via docker-compose. Do I have no other choice than writing myself the config.php file and mount it? I've been looking for something like this

    environment:
      - REDIS_HOST=redis
      - REDIS_PORT=6379

But no example found so far, I guess I can't ?

Best regards,

### PR DESCRIPTION
**Pull request by rodrigoaguilera**
Closes #194

### CODE REVIEW

File: .config/smtp.config.php [hunk 1]
```diff
@@ -0,0 +1,15 @@
+<?php
+if (getenv('MAIL_SMTPMODE') == 'smtp') {
+  $CONFIG = array (
+    'mail_smtpmode' => getenv('MAIL_SMTPMODE'),
+    'mail_smtphost' => getenv('MAIL_SMTPHOST'),
+    'mail_smtpport' => getenv('MAIL_SMTPPORT'),
```
Comment (tilosp) [inline .config/smtp.config.php]: I think that the if statement should check that every var that does not have a default value is set.
For the most values you can simply use the default values of the docs: https://docs.nextcloud.com/server/12/admin_manual/configuration_server/config_sample_php_parameters.html#mail-parameters

File: .config/redis.config.php [hunk 2]
```diff
@@ -0,0 +1,10 @@
+<?php
+if (getenv('REDIS_HOST')) {
+  $CONFIG = array (
+    'memcache.locking' => '\OC\Memcache\Redis',
+    'redis' => array(
+      'host' => getenv('REDIS_HOST'),
+      'port' => getenv('REDIS_PORT') ?: 6379,
+    ),
+  );
+}
```
Comment (tilosp) [inline .config/redis.config.php]: Please add a new line at the end

File: .config/smtp.config.php [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+    'mail_smtpsecure' => getenv('MAIL_SMTPSECURE'),
+    'mail_smtpauth' => getenv('MAIL_SMTPAUTH'),
+    'mail_smtpauthtype' => getenv('MAIL_SMTPAUTHTYPE'),
+    'mail_smtpname' => getenv('MAIL_SMTPNAME'),
+    'mail_smtppassword' => getenv('MAIL_SMTPPASSWORD'),
+    'mail_from_address' => getenv('MAIL_FROM_ADDRESS'),
+    'mail_domain' => getenv('MAIL_DOMAIN'),
+  );
+}
```
Comment (tilosp) [inline .config/smtp.config.php]: Please add a new line at the end

**Comment by tilosp** [comment]
@pierreozoux i think we also need to implement something similar to #180. Because if we don't do this the variables are not available when running Upgrade.

**Comment by tilosp** [comment]
`sudo -u www-data -EH "$@"` should do what we want

File: .config/smtp.config.php (hunk 1 above)
Comment (rodrigoaguilera) [reply .config/smtp.config.php]: Sorry I think I don't quite get what you mean. 
I was thinking to improve it like this:
```php
$CONFIG = array (
  'mail_smtpmode' => getenv('MAIL_SMTPMODE') ?: 'php',
  'mail_smtphost' => getenv('MAIL_SMTPHOST') ?: '127.0.0.1',
  'mail_smtpport' => getenv('MAIL_SMTPPORT') ?: '25',
```
and so on and so on...

Did you refer to that?

**Comment by rodrigoaguilera** [comment]
I do not understand the comment about the sudo

For the readme I was thinking to add a new section after the NEXTCLOUD_DATA_DIR example:

> You can also configure your SMTP server with environment variables:
>   MAIL_SMTPMODE
>   MAIL_SMTPHOST
>   MAIL_SMTPPORT
>   (actually list them all)
> 
> This variables have analogous names as the ones described in the manual 
> https://docs.nextcloud.com/server/13/admin_manual/configuration_server/config_sample_php_parameters.html#mail-parameters
> 

What do you think?

I will change my PR once this points are clarified.

Thank you for the review and such an amazing contribution :)

**Comment by rodrigoaguilera** [comment]
@tilosp Can you clarify so I can continue working on this?

**Comment by tilosp** [comment]
The redis config is fine like this.

The other one should look something like this:
```php
<?php
if (getenv('SMTP_HOST') && getenv('MAIL_FROM_ADDRESS') && getenv('MAIL_DOMAIN')) {
  $CONFIG = array (
    'mail_smtpmode' => 'smtp',
    'mail_smtphost' => getenv('SMTP_HOST'),
    'mail_smtpport' => getenv('SMTP_PORT') ?: (getenv('SMTP_SECURE') ? 465 : 25),
    'mail_smtpsecure' => getenv('SMTP_SECURE') ?: '',
    'mail_smtpauth' => getenv('SMTP_NAME') && getenv('SMTP_PASSWORD'),
    'mail_smtpauthtype' => getenv('SMTP_AUTHTYPE') ?: 'LOGIN',
    'mail_smtpname' => getenv('SMTP_NAME') ?: '',
    'mail_smtppassword' => getenv('SMTP_PASSWORD') ?: '',
    'mail_from_address' => getenv('MAIL_FROM_ADDRESS'),
    'mail_domain' => getenv('MAIL_DOMAIN'),
  );
}
```
(The if statement checks all vars that don't have a default value)

**Comment by unteem** [comment]
@rodrigoaguilera are you still working on this PR? I'm interested in getting this merged as soon as possible. Let me know if you are planning to finish it or if I should make a new PR based on this one and tilosp remarks
thx

**Comment by rodrigoaguilera** [comment]
@unteem 
Please open a new PR and mention me on so I can follow the progress.

Thank you for taking over ;)

**Comment by marceljd** [comment]
@rodrigoaguilera @unteem, what is the current status? For me the SMTP part is not that important, you can set the values through the web interface. Redis cannot be set after spinning up so that would be nice to implement. I tested the Redis part on my setup (Apache 14.03) and it seems So I suggest to remove the SMTP part here and move it to another PR (or leave it). The Redis part only needs one new line and a Readme entry. Could you do that or can you grant me access so I can change this PR?

**Comment by rodrigoaguilera** [comment]
@marceljd Better make a new PR

**Comment by SnowMB** [comment]
Let's close this and use the other PR.

**Pull request closed without merging**